src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/DictionaryModelBinder.cs

```diff
@@ -79,21 +76,26 @@ public class DictionaryModelBinder<TKey, TValue> : CollectionModelBinder<KeyValu
                 // that culture when rendering a form.
                 var convertedKey = ModelBindingHelper.ConvertTo<TKey>(kvp.Key, culture: null);
 
-                valueBindingContext.ModelName = kvp.Value;
+                using (bindingContext.PushContext(
+                    modelMetadata: valueMetadata,
+                    fieldName: bindingContext.FieldName,
+                    modelName: kvp.Value,
+                    model: null))
```

Comment:
As in the `CollectionModelBinder`, can do this outside the loop rather than per iteration.